pandas/core/internals/managers.py

```diff
@@ -384,8 +384,29 @@ def setitem(self, indexer, value) -> Self:
             raise ValueError(f"Cannot set values with ndim > {self.ndim}")
 
         if using_copy_on_write() and not self._has_no_reference(0):
-            # if being referenced -> perform Copy-on-Write and clear the reference
-            # this method is only called if there is a single block -> hardcoded 0
+            # Split blocks to only copy the columns we want to modify, only 1 block
+            if self.ndim == 2 and isinstance(indexer, tuple):
+                blk_loc = self.blklocs[indexer[1]]
+                if is_list_like(blk_loc) and blk_loc.ndim == 2:
+                    blk_loc = np.squeeze(blk_loc)
+                elif not is_list_like(blk_loc):
+                    # Keep dimension and copy data later
+                    blk_loc = [blk_loc]
+                if len(blk_loc.shape) == 0:
+                    return self
```

Comment:
not a shallow copy?

Reply:
IIRC we can only get here in setitem ops, so doesn't matter. Will check more closely though